### PR TITLE
Generate service specific waiter sync and async interfaces

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Generate service specific waiter sync and async interfaces

## Testing
Add codegen tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/docs/WaiterDocs.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    public static String waiterInterfaceJavadoc() {
+        return "Waiter utility class that polls a resource until a desired state is reached or until it is "
+               + "determined that the resource will never enter into the desired state. This can be"
+               + " created using the static {@link #builder()} method";
+    }
+
+    public static CodeBlock waiterOperationJavadoc(ClassName className, Map.Entry<String, WaiterDefinition> waiterDefinition,
+                                                   OperationModel operationModel) {
+        String javadocs = new DocumentationBuilder().description("Polls {@link $T#$N} API until the desired condition "
+                                                                 + "<code>$N</code> is met, "
```
Comment (dagnir) [inline codegen/src/main/java/software/amazon/awssdk/codegen/docs/WaiterDocs.java:39]: nit: `{@code }`

File: codegen/src/main/java/software/amazon/awssdk/codegen/docs/WaiterDocs.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+package software.amazon.awssdk.codegen.docs;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import java.util.Map;
+import software.amazon.awssdk.codegen.model.intermediate.OperationModel;
+import software.amazon.awssdk.codegen.model.service.WaiterDefinition;
+import software.amazon.awssdk.core.waiters.PollingStrategy;
+
+public class WaiterDocs {
```
Comment (dagnir) [inline codegen/src/main/java/software/amazon/awssdk/codegen/docs/WaiterDocs.java:25]: nit: `final`

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
+package software.amazon.awssdk.services;
```
Comment (dagnir) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java:1]: Is this package name correct? Seems like it should be under `software.amazon.awssdk.services.myservice.waiters` or something similar

Comment (zoewangg) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java:1]: Yeah, good call!

**Review by dagnir — APPROVED** [review]

**Pull request merged**